Bluetooth: hci_usb:  Fix bulk read and write related issues

sundar-subramaniyan opened the pull request:
This PR is created to review the hci_usb fixes made in try_write() and btusb_bulk_out()

sundar-subramaniyan committed:
Bluetooth: hci_usb: Write large packets to USB in multiple chunks

try_write() writes the whole net buffer to USB which sometimes is larger
than the maximum packet size of an endpoint.

Use the ret_bytes param to know how much has been written to the USB bus
and write the remaining chunks in the subsequent calls.

Usually the USB driver write doesn't happen immediately in cases where
the usb_write() is not synchronous. In such cases, USB driver returns
-EAGAIN.

If driver returns -EAGAIN, yield so as to give other threads a chance
to run while waiting for previous USB write to complete.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

finikorg commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -661,14 +661,26 @@ DEVICE_INIT(btusb, "btusb", &btusb_init,
+			k_yield();
 			break;
 		/* TODO: Handle other error codes */
 		default:
```
Should you handle other error codes? BTW, could you use similar function try_write() in netusb.c

sundar-subramaniyan replied:
Apart from EAGAIN, we can possibly get EIO, EINVAL, ETIMEDOUT from the current USB drivers in the case of errors.
In the case of other errors I should return instead of checking the wrote value. It should be done only when return value from usb_write is 0.

Also, I believe this patch is only required until usb_transfer() APIs are available in the device stack which would be the correct way to fix this. So for now, I'll handle the length check only if return value is zero. That will be similar to the try_write() implementation in netusb.c

sundar-subramaniyan replied:
One other thing from netusb.c is that they try only for 10 times without yielding. I believe that's going to fail sometimes with certain drivers like the upcoming nRF52840 USB driver. Keeping it an arbitrary value will fail one or the other time. So for now, it's going to be a forever loop if -EAGAIN is returned all the time. If the driver is implemented properly this shouldn't happen though.

sundar-subramaniyan commented:
@finikorg I have reworked on the patch. It is almost similar to what the try_write() in netusb.c does except for the fixed retry count.

sundar-subramaniyan committed:
Bluetooth: hci_usb: Read the whole USB EPOUT buffer at once

btusb_bulk_out() reads the EPOUT buffer in multiple of 4 bytes because
of Quark's USB FIFO size limitation. But this has later been taken care
internally in the DW driver making it redundant in hci_usb.

Remove Quark specific handling and read the whole EPOUT buffer at once.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

carlescufi commented:
CC @pawelzadrozniak

finikorg commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -416,8 +416,13 @@ static void btusb_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
-	u32_t len, read = 0;
+	u32_t len;
+#ifdef CONFIG_SOC_FAMILY_QUARK
+	u32_t read = 0;
```
That is already handled in the usb_dc_ep_read_wait() in usb_dc_dw.c driver. So you can remove CONFIG_SOC_FAMILY_QUARK part.

sundar-subramaniyan replied:
Okay, I'll remove all the Quark specific stuff since you say it's already handled in DW driver.

finikorg commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -451,6 +457,10 @@ static void btusb_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 	}
+#else
+	mem = (u8_t *)net_buf_add(buf, len);
+	usb_read(ep, mem, len, NULL);
```
you do not need to cast to (u8_t *) here. You can also use 
`usb_read(ep, net_buf_add(buf, len), len, NULL);`

sundar-subramaniyan replied:
Okay, that's much simpler. I'll rework and update the PR.

finikorg reviewed (changes requested):
Please remove Quark part, otherwise lgtm.

sundar-subramaniyan commented:
@finikorg I have addressed your comments. Please take a look.
